internal/errors.go

```diff
@@ -2,46 +2,186 @@ package internal
 
 import (
 	"bytes"
-	"errors"
 	"fmt"
-
-	"github.com/cilium/ebpf/internal/unix"
+	"io"
+	"strings"
 )
 
-// ErrorWithLog returns an error that includes logs from the
-// kernel verifier.
+// ErrorWithLog returns an error which includes logs from the kernel verifier.
+//
+// The default error output is a summary of the full log. The latter can be
+// accessed via VerifierError.Log or by formatting the error, see Format.
 //
-// logErr should be the error returned by the syscall that generated
-// the log. It is used to check for truncation of the output.
-func ErrorWithLog(err error, log []byte, logErr error) error {
+// A set of heuristics is used to determine whether the log has been truncated.
+func ErrorWithLog(err error, log []byte) *VerifierError {
+	const whitespace = "\t\r\v\n "
+
 	// Convert verifier log C string by truncating it on the first 0 byte
 	// and trimming trailing whitespace before interpreting as a Go string.
+	truncated := false
 	if i := bytes.IndexByte(log, 0); i != -1 {
+		if i == len(log)-1 && !bytes.HasSuffix(log[:i], []byte{'\n'}) {
+			// The null byte is at the end of the buffer and it's not preceded
+			// by a newline character. Most likely the buffer was too short.
+			truncated = true
+		}
+
 		log = log[:i]
+	} else if len(log) > 0 {
+		// No null byte? Dodgy!
+		truncated = true
 	}
-	logStr := string(bytes.Trim(log, "\t\r\n "))
 
-	if errors.Is(logErr, unix.ENOSPC) {
-		logStr += " (truncated...)"
+	log = bytes.Trim(log, whitespace)
+	logLines := bytes.Split(log, []byte{'\n'})
+	lines := make([]string, 0, len(logLines))
+	for _, line := range logLines {
+		// Don't remove leading white space on individual lines. We rely on it
+		// when outputting logs.
+		lines = append(lines, string(bytes.TrimRight(line, whitespace)))
 	}
 
-	return &VerifierError{err, logStr}
+	return &VerifierError{err, lines, truncated}
 }
 
 // VerifierError includes information from the eBPF verifier.
+//
+// It summarises the log output, see Format if you want to output the full contents.
 type VerifierError struct {
-	cause error
-	log   string
+	// The error which caused this error.
+	Cause error
+	// The verifier output split into lines.
+	Log []string
+	// Whether the log output is truncated, based on several heuristics.
+	Truncated bool
 }
 
 func (le *VerifierError) Unwrap() error {
-	return le.cause
+	return le.Cause
 }
 
 func (le *VerifierError) Error() string {
-	if le.log == "" {
-		return le.cause.Error()
+	log := le.Log
+	if n := len(log); n > 0 && strings.HasPrefix(log[n-1], "processed ") {
+		// Get rid of "processed 39 insns (limit 1000000) ..." from summary.
+		log = log[:n-1]
+	}
+
+	output := 0
+	switch n := len(log); n {
+	case 0:
+		return le.Cause.Error()
+
+	case 1, 2:
```

Comment:
I've pushed a commit that changes this. Feel free to remove it if there was a strong reason for `case 1, 2`.

Reply:
(culprit is likely torvalds/linux@7df5072cc05fd1aab5823bbc465d033cd292fca8)

Reply:
If I understand correctly, we'd always output `R0 !read_ok` now. Arguably, `0: (95) exit: R0 !read_ok` is a better error message though, since it conveys that there is no valid return value. Maybe we should instead always output the last two lines? Or include another line if the last line starts with `R\d+`? If this turns out to be too complicated we can get rid of the 1-2 line thing though.

Also, could you add two new test cases based on the snippets you posted? Since this is a heuristic additional cases will be super helpful when we have to tweak stuff in the future.

Reply:
> Arguably, `0: (95) exit: R0 !read_ok` is a better error message though ... Or include another line if the last line starts with R\d+?

Does the preceding line always pertain to the actual instruction that caused the invalid access? That could be helpful, I guess. Still missing a lot of context, ideally you'd want the corresponding btf lineinfo, but we need to keep it short for `%s`.

Seems easy enough, we could just add another condition that expands the line range by one? Since we know we're on the next-to-last line, a simple `strings.HasPrefix(log[n-1], "R")` could be sufficient. 

WDYT?

Reply:
This could just work:
```
λ  linux master ✓ grep -E "verbose\(env, \"R" kernel/bpf/verifier.c | wc -l
60
```

```
λ  linux master ✓ grep -E "verbose\(env, \"R[^%]" kernel/bpf/verifier.c
			verbose(env, "R0 not a scalar value\n");
		verbose(env, "R0 leaks addr as return value\n");
```

All verbose statements that start with R are currently used to describe a register.

Slippery slope, though. The more complex the heuristic, the deeper the rabbit hole goes and the more things we'll have to account for eventually. I don't think that's what we want to do. We should probably limit it to the last line (bar `processed ...`) and encourage the use of `errors.As` for anything more advanced. For users deploying onto infrastructure they don't control, this is kind of a must.

Reply:
I think the heuristic is worth a shot. If it doesn't pan out we can remove it. I've pushed a commit which adds it.

Removing the special case for 2 lines in the log makes sense.